enhancements/network/ovn-multi-external-gateway.md

```diff
@@ -0,0 +1,183 @@
+
+## Proposal
+
+Create a new CRD that contains a direct relationship to the annotations used in the pod/namespace to manage the external gateways configuration:
+
+Namespace annotations:
+* `k8s.ovn.org/routing-external-gws`
+* `k8s.ovn.org/external-gw-pod-ips`
+
+Pod annotations:
+* `k8s.ovn.org/routing-namespaces`
+* `k8s.ovn.org/routing-network`
+* `k8s.ovn.org/bfd-enabled`
+
+The implementation will also include a new controller that will watch cluster wide for instances of this CRD and apply the changes to the OVN network configuration as it is currently done by the annotation based logic.
+
+The controller will use the OpenShift release version as well as the existence of the annotations in the namespace and pods to apply the changes to the network configuration:
+
+| **OpenShift version** | **Annotations exist** | **CR exists** | **Action**                                        |
+|-----------------------|-----------------------|---------------|---------------------------------------------------|
+| 4.13                  | No                    | Yes           | Use CR information to configure external gateways |
+| 4.13                  | Yes                   | Yes           | Use annotations to configure external gateways    |
+| 4.13                  | Yes                   | No            | Use annotations to configure external gateway     |
+| 4.13                  | No                    | No            | Nothing to do                                     |
+| 4.14                  | No                    | Yes           | Use CR information to configure external gateways |
+| 4.14                  | Yes                   | Yes           | Use CR information to configure external gateways |
+| 4.14                  | Yes                   | No            | Use annotations to configure external gateways    |
+| 4.14                  | No                    | No            | Nothing to do                                     |
+
+The current annotation based logic will be enhanced as part of this change to support the logic described in the previous table.
+
+### Workflow Description
+The workflow follows the behavior of a namespaced CR.
+
+
+### API Extensions
+
+A new namespaced-scoped CRD is created to represent the multiple external gateways configuration. The annotations are mapped as following:
+* `k8s.ovn.org/routing-external-gws` is a namespace specific annotation and is represented in the CRD as `routingExternalGateways`. The values represent a slice of strings containing the IPs used for external traffic. 
+This annotation works in conjunction with `k8s.ovn.org/external-gw-pod-ips` and is used to capture the IPs that are stored in conntrack.
+* `k8s.ovn.org/routing-namespaces` is mapped as `routingNamespaces` and it also contains a slice of the namespaces linked to this gateway configuration.
```

Comment:
we can drop this because the CRD is namespace scoped. So it is implied that this CRD would affect pods in the same namespace. Alternatively, we could make the CRD global and then keep this field. The advantage would be one one CR to serve multiple namespaces, but then that would mean less granular RBAC between namespace traffic. I'm kind of inclined to keep it namespace scoped, but open to other's opinions.

Reply:
My understanding is that the current implementation uses this annotation to allow gw pods to accept traffic from multiple namespaces. I'm fine with namespace scoped and remove this field as you suggested, but I worry if doing so we would be not providing the same functionality that exists with annotations. Thoughts?

Reply:
Right, the user would have to make several CRDs, one for each namespace they want to route traffic from. That is a bit different than today as they can declare in a single place they want traffic from all namespaces. That seems less secure though, which is why I was thinking namespaced scope would be better.

@fedepaol wdyt?

Reply:
Talking more with @jordigilh I think there is another angle we need to consider. If the object is namespace scoped then a tenant in that namespace would be able to control how their traffic egresses. In the real world use case I would imagine cluster admins want to control the traffic of their tenants and do not want to give this kind of access out. So maybe that is a good reason to make it cluster scoped and require cluster admin RBAC.

Reply:
So talking in the OVNK community call yesterday with, Nvidia showed us their downstream policy based routing CRD, which has some overlap with what we are trying to do. So rather than having 2 different CRDs for routing, we discussed unifying into a single CRD. However, we would only implement pieces that we cared about (multiple external gateways) and Nvidia would bring their corresponding implementation for other cluster routing upstream later.

Example of their CRD:
```yaml
apiVersion: k8s.ovn.org/v1beta1
kind: AdminPolicyBasedRoute
metadata:
  name: honeypotting
spec:
  networkAttachmentName: default/ovn-primary
  policies:
  - from:
      podSelector:
        matchLabels:
          under_attack: “true”
    to:
      cidr: 0.0.0.0/0
    nexthop:
      ips:
      - 10.192.14.9
```

Nvidia is using this to match certain traffic to a particular destination, and set the next hop. They do this by Logical Route Policy in OVN on ovn_cluster_router, while we are using Logical Router Static Routes in the gateway router (GR). The reason for using logical route policy is they can match on both source and dest in a route (as well as potentially other criteria), while if they were to use only source or dest they could use regular static routes. However, policy routing takes precedence over static routing in OVN, so we need to be careful here that no policy should ever overlap if we mix policy and static routes.

Here is my first stab at something that we could use as unified CRD for both cases (assume cluster CIDR of 10.244.0.0/24, external network of 172.18.0.0/24):

```yaml
apiVersion: k8s.ovn.org/v1beta1
kind: AdminPolicyBasedRoute
metadata:
  name: honeypotting
spec:
  policies:
  ##intra cluster example
-  trafficType: internal
   from:
      podSelector:
        matchLabels:
          under_attack: “true”
    to:
      cidr: 9.0.0.0/24
    nexthops:
      - ip: "10.244.0.5"
        bfdEnabled: false
      - ip: "10.244.0.6"
        bfdEnabled: false
      - host
    networkAttachmentName: default/ovn-primary

  ## gateway example
- trafficType: external
  from:
    namespaceSelector:
       matchLabels:
          multiple_gws: true
    nexthops:
      - ip: "172.18.0.2"
        bfdEnabled: true
      - ip: "10.244.0.3"
        bfdEnabled: false
      - podSelector:
          matchLabels:
            external-gateway: true
        bfdEnabled: true
        networkAttachmentName: sriov1

status:
  externalGatewayIPs: ["172.18.0.2", "172.18.0.3", "172.18.0.4", "172.18.0.5"]
  internalGatewayIPs: ["10.244.0.5", "10.244.0.6"]
```

From the above we get can define a few behaviors (a few of these ideas are very long term, but just thinking outside the box):

1. CRD is cluster scoped for security.
2. If traffic type is internal, the rules are applied to ovn_cluster_router. If external, then applied to GR.
3. If a policy consists of both from and to, then logical route policy will be used. If only source or dest are provided, then logical static routes will be used.
4. Policies should not overlap.
5. In the first example there are 3 nexthops provided. Two are static ip addresses inside the cluster subnet, while a third nexthop "host" will indicate to route packets inside the local host where the pod lives (aka mgmt port). This allows users to redirect traffic to the kernel in case they want to use the host ip routing to egress a different interface.
6. networkAttachmentName defines either the network of the selected pods (primary or secondary) to apply policy to or choose IP from a pod acting as an external gateway.
7. networkAttachmentName is not applicable for trafficType external on selected pods to apply policy to as secondary networks have no ingress/egress.

Reply:
I am not sure the status above makes sense, or at least we need to provide a status per namespace.
Not sure also that the internal / external naming is appropriate here as (iiuc) even the internal one might affect egress traffic.

Re - the host bit. It'd be nice if we could specify the next hop _on the host_. That would basically address what we are trying to achieve with services.

Reply:
Two things here, 
1) I feel the MEG CRD should be cluster-scoped, not namespace scoped - a namespace scope doesn't make sense here; the feature is more for admins not users, similar to egressIPs...

2) I am slightly skeptical about merging this CR with what Nvidia is doing for another feature? Reason being we can't really change/expand CRDs like that over releases right, they are API breaking changes? Its going to be hard to get all the fields right in such a way that we can only expand in the future and not break the type of fields..., maybe we keep this v1alpha1 and hoping this feature is dev preview until Nvidia also brings its things in. OR I am missing details and maybe we have a doc from Nvidia that explains its feature in depth so that we are covered? -> example is their feature also for egress traffic or is it for internal traffic? Mixing those things up into a single CRD could get confusing if these two features are doing opposite things on a high level.

3rd extra point which is food for thought - how would the controller know whether its an internal v/s external gw thing here depending on which we'd need to do things on GR v/s ovn_cluster_router? - we'd need that signal somehow from the CRD i guess..

Reply:
> Two things here,
> 
>     1. I feel the MEG CRD should be cluster-scoped, not namespace scoped - a namespace scope doesn't make sense here; the feature is more for admins not users, similar to egressIPs...

Yes, the CRD is cluster scoped. My initial proposal had a mix of cluster and namespaced but that's no longer the case.

> 
>     2. I am slightly skeptical about merging this CR with what Nvidia is doing for another feature? Reason being we can't really change/expand CRDs like that over releases right, they are API breaking changes? Its going to be hard to get all the fields right in such a way that we can only expand in the future and not break the type of fields..., maybe we keep this v1alpha1 and hoping this feature is dev preview until Nvidia also brings its things in. OR I am missing details and maybe we have a doc from Nvidia that explains its feature in depth so that we are covered? -> example is their feature also for egress traffic or is it for internal traffic? Mixing those things up into a single CRD could get confusing if these two features are doing opposite things on a high level.

What I'm proposing is to use 2 CRDs, one for internal and another one for external traffic. We can keep the CRD structures similar but focus on what matters to us. 

> 
> 
> 3rd extra point which is food for thought - how would the controller know whether its an internal v/s external gw thing here depending on which we'd need to do things on GR v/s ovn_cluster_router? - we'd need that signal somehow from the CRD i guess..

See my comment above.